src/applications/apps.js

```diff
@@ -50,12 +50,18 @@ export function registerApplication(appName, arg1, arg2) {
   if (typeof activeWhen !== 'function')
     throw new Error(`The activeWhen argument must be a function`);
 
-  apps.push({
-    name: appName,
-    loadImpl,
-    activeWhen,
-    status: NOT_LOADED,
-  });
+  const app = {
+      name: appName,
+      loadImpl,
+      activeWhen,
+      status: NOT_LOADED,
+  };
+
+  if (customProps) {
```

Comment:
Would you consider adding this new argument as a function to give added functionality?

So you would do something like
```
declareChildApplication(
    'example',
    () => import(...),
    () => true,
    (app, apps) => ({ ...app, appCount: apps.length })
)
```

Reply:
Hi, thank you for your comment. Though I don't know what you actually mean. My PR enables you to pass **anything** down from the root application to your child application (including functions). Have a look at my examples above (which I just added).

Reply:
The important piece is that the function you pass in gets **invoked** with additional information.

For example you could go a step further and do
```js
if (isFunction(customProps)) {
    app.customProps = customProps(app, apps, getMountedApps(), getAppsToMount());
}
```

Reply:
Alright, now I get what you mean. The problem is, that this would change **every** function I would want to pass down to my child apps. 

Here is a made up example where I want to pass down a function to my child app that lets my child app access some globaly stored accessTokens for various APIs:

```
// root app

const getAccessToken = (identifier) => {
    const accessTokens = { 
        googleAPI: 'jkd7fd8723h82387z4g283bzn4', // accessTokens may be dynamic and change over time
        mongoDB: 'sd99j82398jf982j9823j'
    };
    return accessTokens[identifier];
};

registerApplication (
    'childApp1',
    () => import(...),
    () => true,
    (identifier) => getAccessToken(identifier)
)
```

```
// child app1
export function bootstrap(props) {
    const getAccessTokenFn = props.customProps; // I could store this somewhere
    console.log(getAccessTokenFn('googleAPI')); // And whenever I need an access token,												
                                                // I could call the function which is defined in the root app
    return ngLifecycles.bootstrap(props);
}
```

Now with your suggestion:
```
if (isFunction(customProps)) {
    app.customProps = customProps(app, apps, getMountedApps(), getAppsToMount());
}
```

This would totally break. Also I think it is kind of strange that for some reason functions get overwritten by the framework. But anything else is passed down correctly.


**BUT** I think your idea definitely has use cases. And it is actually totally easy to accomplish something like that if you just add down anything of interest inside the lifecyle methods within the framework. Just like:

```
// bootstrap.js
await reasonableTime(app.bootstrap({appName: app.name, app: app }), `Bootstrapping app '${app.name}'`, app.timeouts.bootstrap); // Here I pass app down to the child application
```

Reply:
Think perhaps we're still not on the same page.

Was thinking something along the lines of this
```js
export function registerApplication(appName, arg1, arg2, transformProps) {
 ...

 let app = {
      name: appName,
      loadImpl,
      activeWhen,
      status: NOT_LOADED,
  };

  if (isFunction(transformProps)) {
      app = transformProps(app, apps);
  }

  apps.push(app);
```

Ultimately your PR solves my simple use case, so am fine to continue iterating on this after pushing this PR through. Any other thoughts @mbanting @joeldenning?

Reply:
Good discussion here, and apologies for the delayed response -- I would like to understand some of the use cases for this better.

What's an example use case where you want the custom props to be based on the current app and all the rest of the apps?

Reply:
Just had an idea that this may be a good place to inject additional information that we don't have access to like
```js
app = transformProps(app, apps, getMountedApps(), ...)
```
I don't necessarily have a true use case in mind, but passing in a function would give us more flexibility going forward. Though the more I write it seems like I'm searching for a problem for my solution 😄